iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -219,8 +223,8 @@ await _internalRetryPolicy
                             // Ensure that the connection has been opened, before enabling the callback for receiving messages.
                             await EnsureOpenedAsync(false, cancellationToken).ConfigureAwait(false);
 
-                            // Wait to acquire the _handlerSemaphore. This ensures that concurrently invoked API calls are invoked in a thread-safe manner.
-                            await _handlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                            // Wait to acquire the _cloudToDeviceSubscriptionSemaphore. This ensures that concurrently invoked API calls are invoked in a thread-safe manner.
+                            await _cloudToDeviceMessageSubscriptionSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Why null check on 237 but not here? Do we ever expect this to be null?

Reply:
Dispose() will set the semaphores to null and I don't expect this code to be hit asfter Dispose() has been called.

Reply:
I don't love the assumption, but I'll leave it to you.